File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length, max = 0;

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i <= n; i++){
            while (!stack.isEmpty() && (i ==n || heights[i] < heights[stack.peek()])){
                int height = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - 1 - stack.peek();
                max = Math.max(max, width * height);
            }
            stack.push(i);
        }

        return max;
    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area, n = 0, len(heights)
        stk = []

        for i in range(n + 1):
            while stk and (i == n or heights[i] < heights[stk[-1]]):
                height = heights[stk.pop()]
                width = i - 1 - stk[-1] if stk else i
                max_area = max(max_area, height * width)          
            stk.append(i)

        return max_area


        
        